### DISCUSSION TITLE
[Checklist] How to make sure apiVersion migration is comprehensive

### DISCUSSION THREAD

**Discussion by JeromeJu** [tektoncd/pipeline#5800] [General]
This is to keep tracks of the issues during the process of migration from v1beta1 to v1. And this aims to spot and avoid the "gotchas" alongside each stage. And hopefully this would also note down the difference between the last migration from `v1alpha1` to `v1beta1` and have some sort of checklist for future migrations.

## Checklist

Stage of creating CRDs of newer apiVersion:
- [ ] Reach consensus of deprecation of fields and the respective handling of them in both apiVersions

Stage of serving new version of CRD:
- [ ] Integration tests for conversion
- [ ] Example tests with the new clients
- [ ] Conversion unit tests covering all fields

------
- How to catch the issue of `STATUS`:
  - Thoroughly testing:
  The examples_tests needs to be updated for v1clients
- Conversion of bundles fixes:
  - Integration tests needs to be introduced before serving 

Gotchas:
Kubernetes always present the highest priority with kubectl at
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/

Diff:
- Conversion of `STATUS` of PR and TR in `v1alpha1` to `v1beta1`, the status are shared (just copied from v1)

**Comment by JeromeJu** [comment]
[WIP] Comparing the process for `v1beta1 -> v1` and `v1alpha1 -> v1beta1`

`v1beta1 -> v1`:

- Add CRD Golang structs
  - #5125   
  - #5219 
  - #5264
  - #5324  
- Add conversion
  - #5202
  - #5258
  - #5274 
  - #5426 
- Handle deprecated fields in conversion
  - #5253
  - #5331 
  - #5378
  - #5381  
  - #5477 
- Add CRD to webhook and config
  - #5234 
  - #5464 
  - #5379 
  - #5464 
- Add migration doc
  - #5415 
- Release CRD
  - #5579  

Comparing with what's done for `v1alpha1 -> v1beta1`:
- https://github.com/tektoncd/pipeline/pull/2577

**Comment by JeromeJu** [comment]
Post-mortem:
- [ ] Compare the Cutoff of V1 CRD release
- https://github.com/tektoncd/pipeline/pull/5974
- https://github.com/tektoncd/pipeline/pull/5948

 
- v1 software release:
- Comparing with [v1alpha1->v1beta1 swap](https://github.com/tektoncd/pipeline/pull/2577), we have experienced more obstacles/blockers while switching from v1beta1->v1
  - we didn't have feature flags at that time, so we didn't need to deal with https://github.com/tektoncd/pipeline/issues/6592.
  - we had introduced features/ flags that needs to be cleanup before migrating
    -  pipelineRunStatus Migration
    - pipelineResources Migration
    - version validation (coupling of apiVersion and feature version)

**Comment by dibyom** [comment]
Nice @JeromeJu we should update https://github.com/tektoncd/pipeline/blob/main/docs/developers/api-versioning.md#adding-a-new-api-version-to-a-pipelines-crd with this when done!
